Add convenience macro for declaring types

abadams opened the pull request:
Helps with name mangling in code using define_extern that passes
pointers to user types as arguments.

zvookin commented on src/Type.h:
```diff
@@ -118,32 +118,45 @@ struct halide_c_type_to_name {
+HALIDE_DECLARE_EXTERN_STRUCT_TYPE(buffer_t);
+HALIDE_DECLARE_EXTERN_STRUCT_TYPE(halide_filter_metadata_t);
+
+// You can make arbitrary user-defined types be "Known" using the
```
This should document that the macro invocation must be done in global scope.

zvookin commented:
LGTM.

steven-johnson commented:
LGTM

abadams commented:
I discovered the forward declaration code in CodeGen_C was messed up, so I fixed that up and added support for unions too, and also added some testing. This PR has thus substantially changed. Please take another look.
